### PR TITLE
Coords

### LINKED ISSUE #25: Use {0, 1, 2} globally for Piece position.
We sometimes use -1, 0, 1; and sometimes 0, 1, 2. We need to be consistent or bugs will occur.

### LINKED ISSUE #30: Add piece animations
Tracking bug for adding animations so pieces actually move between turns.

As a side note, this should make it way clearer what actually happened each turn :-)

### LINKED ISSUE #33: Animations of overlapping pieces of the same player transmorgify one into just a direction
If a player has two pieces converging on the same spot, one of them gets turned into an image of just its direction during the animation. This is because that's exactly what Space.updateImageResources is supposed to do.

### PR DESCRIPTION
**Pull request by apmcleod**
Animations are working now I think! That includes wraparounds, halfway collisions, and full collisions. There is a slight pause halfway while it figures out all of the collisions, but it doesn't look too bad (actually, I kind of prefer it like that).

This also fixes #25, as I also changed the Pieces to use [0,1,2] for position, and changed things to use "row" "column" "horizontal" and "vertical" instead of x and y.

This fixes #33, as I call Piece.updateImageResourceFullPiece() before animating.

This fixes #30 as well. This is how each the animation logic works:
1. Create and add any dummy Pieces. These Pieces are initially off the edge of the board, and are used to ensure that wrap-arounds animate properly.
2. Animate all Pieces (including dummies) halfway.
3. Calculate and resolve halfway collisions (We could move the calculation out of this step to a step 0, but it doesn't seem too bad to me). This also performs any resolution on each Piece's dummies (player changes or removals).
4. Animate all Pieces (including dummies) the rest of the way.
5. Calculate and resolve full collisions (including for dummies a la step 3).
6. Remove all views from the board and repopulate. This both removes any dummy Pieces, as well as ensures that the normal Pieces are now moved to their proper locations (in case they were animated off the edge of the Board.

As a note, this also effects #26, since I removed redo/undo to make things simpler. If we add it back, we should look at doing it in a more efficient way, saving only what we need to.

### CODE REVIEW

File: src/com/tictacgo/data/Board.java [hunk 1]
```diff
@@ -422,102 +390,73 @@ private boolean resolveCollision(List<Piece> collision) {
...
     }
 
-    public void updateUiPositions() {
-        // Iterate through the board and update each piece's UI position.
-        for (List<Space> row : spaces) {
-            for (Space space : row) {
-                space.updateUiPosition();
-            }
+    /**
+     * Update and get the dummy Pieces for each active Piece.
```
Comment (jimmyfnugent) [inline src/com/tictacgo/data/Board.java]: Some more explanation here would be good -- what are dummy pieces and why do we need them?

File: src/com/tictacgo/data/Piece.java [hunk 2]
```diff
@@ -1,32 +1,50 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
 /**
  * A Piece represents a single game piece (X or O), and information about its location, direction,
  * as well as graphical information from the ImageView class.
  */
 public class Piece extends ImageView {
 
+    private final Drawable X_FULL_DRAWABLE = ContextCompat.getDrawable(getContext(),
```
Comment (jimmyfnugent) [inline src/com/tictacgo/data/Piece.java:27]: Nit: these would be better as static

File: src/com/tictacgo/data/Piece.java [hunk 3]
```diff
@@ -235,20 +240,113 @@ public boolean isO() {
...
+            setImageDrawable(O_DIR_DRAWABLE);
+        }
+    }
+
+    /**
+     * Get a List of the dummy Pieces we will need for wrap-animation.
+     *
+     * @return A List of dummy Pieces for wrapping around the edges of the Board.
+     */
+    public void updateDummyPieces() {
```
Comment (jimmyfnugent) [inline src/com/tictacgo/data/Piece.java:265]: IMO this class should be a fairly dumb "model" object that doesn't have any complicated logic in it. I'd move these new methods somewhere else like Board.

**Review by jimmyfnugent — CHANGES_REQUESTED** [review]
Overall, looks really good! Thanks for doing this :-) got a few small comments,

File: src/com/tictacgo/data/Piece.java (hunk 2 above)
Comment (apmcleod) [reply src/com/tictacgo/data/Piece.java:27]: For some reason, this requires a Context, which cannot be gotten statically. I agree that I'd rather not have 4*numPieces of these Drawables though... It seems there should be some possible optimization though, given that all of the Pieces are the same size.

File: src/com/tictacgo/data/Board.java [hunk 4]
```diff
@@ -422,102 +390,73 @@ private boolean resolveCollision(List<Piece> collision) {
...
+     *
+     * @return The dummy Pieces for each active Piece in a List.
+     */
+    public List<Piece> getDummyPieces() {
+        List<Piece> dummies = new ArrayList<>();
+
+        for (Piece piece : pieces) {
+            piece.updateDummyPieces();
+            dummies.addAll(piece.getDummies());
         }
```
Comment (apmcleod) [inline src/com/tictacgo/data/Board.java:417]: So maybe here, replace the inside of the loop with: dummies.addAll(getDummyPieces(piece), and move the logic from Piece to the new method?

File: src/com/tictacgo/data/Board.java [hunk 5]
```diff
@@ -422,102 +390,73 @@ private boolean resolveCollision(List<Piece> collision) {
...
-                posy = 0;
+    public Animator getHalfwayAnimator() {
+        AnimatorSet animator = new AnimatorSet();
+
+        List<Animator> pieceAnimators = new ArrayList<>(pieces.size());
+        for (Piece piece : pieces) {
+            piece.updateHalfwayAnimator();
+            piece.updateImageResourceFullPiece();
+            pieceAnimators.add(piece.getHalfwayAnimator());
         }
```
Comment (apmcleod) [inline src/com/tictacgo/data/Board.java:435]: And here, change the inside of this loop to:

piece.updateImageResourceFullPiece();
pieceAnimators.add(getHalfwayAnimator(piece);

and move the logic from Piece to the new method here?

File: src/com/tictacgo/data/Piece.java (hunk 3 above)
Comment (apmcleod) [reply src/com/tictacgo/data/Piece.java:265]: That makes sense, so something similar to my two other comments?

Comment (jimmyfnugent) [reply src/com/tictacgo/data/Piece.java:265]: Yup! Exactly what I was thinking.

File: src/com/tictacgo/data/Piece.java (hunk 2 above)
Comment (jimmyfnugent) [reply src/com/tictacgo/data/Piece.java:27]: Oh, I see. This is fine then.

File: src/com/tictacgo/data/Board.java [hunk 6]
```diff
@@ -422,102 +390,73 @@ private boolean resolveCollision(List<Piece> collision) {
...
-                                           //And the Array goes from 0 to 2
-        spaces.get(piece.getXPosition() + 1).get(piece.getYPosition() + 1).removePiece(piece);
+    public void removePiece(Piece piece) {
+        spaces.get(piece.getRow()).get(piece.getColumn()).removePiece(piece);
         pieces.remove(piece);
-        ((ViewManager)piece.getParent()).removeView(piece);
+
+        piece.setVisibility(View.GONE);
+        for (Piece dummy : piece.getDummies()) {
+            dummy.setVisibility(View.GONE);
```
Comment (apmcleod) [inline src/com/tictacgo/data/Board.java:399]: This is the reason that the dummies (as the logic is now) MUST be tied to each one's Piece specifically. They are animated together in Piece.getHalfwayAnimator(), and they are hidden together here. I'm sure we could refactor things around, but we need to be able to get from a Piece to its dummies, and the most natural way to do that is to compute and store the dummies within each Piece itself.

File: src/com/tictacgo/data/Piece.java (hunk 3 above)
Comment (apmcleod) [reply src/com/tictacgo/data/Piece.java:265]: See my comments elsewhere. I tried this, but it turns out that Board needs to be able to get from a Piece to its dummies for dummy removal halfway through animation.

**Comment by apmcleod** [comment]
I expanded the dummy comment as requested. Also, I don't think it's feasible at this time to move the dummy logic into Board. Specifically, the way the animation is done, there needs to be a good way to get from a Piece to its dummies directly. If animations (including halfway explosions) are precomputed at some point, we can look back at this.

Other than that, I think I've addressed all of your comments.

**Comment by jimmyfnugent** [comment]
Okay, sounds good. Let's revisit the refactoring later. Thanks for doing this!

**Review by jimmyfnugent — APPROVED** [review]

**Pull request merged**